### PR TITLE
Version number has been added to the UI

### LINKED ISSUE #77: A version number needs to exist on the UI screen for bug reports
It would be nice to see which version is running so that we can ensure that users can report the version that they are running without having to look in the code.

### PR DESCRIPTION
**Pull request by shawnmjones**
I have not found another way to incorporate the Git revision into the docker build. Fixes #77.

### CODE REVIEW

**Comment by ibnesayeed** [comment]
I do not like the functionality of the code being dependent on the entire `.git` history (which should never land into the production environment). I also don't think we even need this level of detail just for the sake of bug reporting. I would perhaps have a periodic release in which case the version number will be part of the package metadata, hence, easier to report anywhere as needed. I would expect people would be running those periodic releases (with manually crafted version numbers). Those who can run the bleeding edge version from the source will perhaps know how to report it.

File: .dockerignore [hunk 1]
```diff
@@ -1,5 +1,5 @@
-.git
 .gitignore
 README.md
 Dockerfile
 LICENSE
+CONTRIBUTING.md
```
[the remark this answers is not in the document]
Comment (shawnmjones) [reply .dockerignore]: I'll see if vscode has it. I'm surprised it isn't the default.

[the remark this answers is not in the document]
Comment (ibnesayeed) [reply .dockerignore]: In VSCode you can set `"files.insertFinalNewline": true` for this effect.

**Comment by ibnesayeed** [comment]
If you are really after having the commit hash, I think I have a mechanism which would involve git hooks. I have not seen it anywhere in practice for this purpose, but it might work. Basically, we can write a git hook script that extracts the latest revision number and write it to an **untracked** file, say, `.headrev`. We can associate this script on various hooks that might make changes to the head revision (for example clone, checkout, pull, and commit etc.). In the version reporting function, we can use manually assigned VERSION constant and append the content of `.headrev` file if present, otherwise silently ignore the last part.

**Comment by shawnmjones** [comment]
I'll continue to use your example of `.headrev` as an untracked file. It is similar to what we used to do when I was a build manager.

I've looked at git hooks. Based on my understanding, they are local. One can check in some scripts and then install them locally, but otherwise they do not travel with the repository.

From https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks:


Based on your suggestion, I thought I would just create a hook that would generate the `.headrev` file after some event. Unfortunately, one still needs the repository (the `.git` directory) to be present in the Docker environment when performing these operations. As I understand it, the Docker build only follows instructions in the `Dockerfile`. I see no other configuration options for the build on DockerHub.

Then I looked into GitHub Webhooks: https://developer.github.com/webhooks/creating/. These sent a POST to a specific URI when certain actions happen in the repository. While useful in other cases, these are not useful in this case. Again, the file must be present in the repository that DockerHub clones and the only way to generate the revision within the `Dockerfile` is by having the `.git` directory present.

There are those who have successfully embedded the revision into the resulting Docker image, but not as part of a DockerHub build. [This guy uses environment variables for his local builds.](https://blog.scottlowe.org/2017/11/08/how-tag-docker-images-git-commit-information/). [This was also a feature request for Moby at some point](https://github.com/moby/moby/issues/14191). [This has also been requested of DockerHub](https://github.com/docker/hub-feedback/issues/600).

So, it appears that one can use DockerHub builds that (temporarily) contain the `.git` directory for this purpose, or one can push their own builds from their desktop to DockerHub. I do not want to give up the DockerHub builds because, as we have already seen, they provide an independent environment that can expose issues.

I wonder if I could use multi-stage builds: https://docs.docker.com/develop/develop-images/multistage-build/

**Comment by ibnesayeed** [comment]
I forgot the fact that git hooks are not tracked and rightfully so, because they will be major security threat if someone clones a repo and suddenly a malicious script gets executed. The proposed approach is still valid. However, it would need some additional effort and more importantly, it will not work in DockerHub.

Multi-stage builds can be used to avoid packing unnecessary history in the final image. However, it does require the inclusion of the `.git` directory during the build phase.

My personal opinion would be to just rely on handcrafted version numbers and release them often. This way, we do not need to be too precise about the specific commits in the reported version numbers. If its DockerHub builds are configured the same way IPWB and MemGator's builds are then people will mostly be using the latest released version and not the bleeding edge version of the master branch, so their version number will precisely reflect the state of the code. Working too hard to somehow bring the commit hash in version number is too much effort for a very little gain.

**Comment by shawnmjones** [comment]
If we were working in a different environment, I would disagree with you. In past projects, we inserted the subversion revision number into the released product. This allowed us to spot process problems and saved us time tracking down issues because we could identify if a fix had been included in a later revision. It also allowed us to identify when build components were not being integrated properly.

That being said, in this environment, I control everything and the project is not that complex. If I have to make a tag with a new version number after every pull request, then I can manually do that. For consistency, it will require ensuring that the version in the `setup.py` matches the tag. I was just trying to avoid the manual effort and let the system work do the work for me.

**Comment by ibnesayeed** [comment]
If the project was a popular one with many users and contributors, efforts to bring such precision in version number would be justified. I still think people who can checkout the master branch and run it from the source should do it for testing only and not for production. Also, they should be able to find their way to report the version number for investigation, if necessary. For masses, a periodically released version is the way to go, which is not a lot of effort for a project like this. Making people run released versions has this advantage of them not running the software with half-baked features. master branch has no guarantee to be in functional state all the time.

Adding a lot of code and moving parts just to ensure inclusion of the revision hash does not add any significant feature towards the actual purpose of the tool, while it makes it more difficult to maintain. By introducing more dependencies, we are introducing more complexities. For example, someone with the tarball of the code (that generally does not include the history) will not be able to build the code, and if we add revision has to be optional then the builds will not be consistent.

Keeping the tag consistent with the version string is not a difficult thing either. However, this can be scripted easily. For example, in the IPWB repo we have a [release.sh](https://github.com/oduwsdl/ipwb/blob/master/release.sh) script which we periodically run. This script changes the version number, created a commit, and tags the repo, then pushes it to GitHub as well as a release on PyPi at the same time. The script uses a [CalVer](http://calver.org/) style, but you can use [SemVer](https://semver.org/) with slight modification where you take the version number as an argument to the script.

**Comment by shawnmjones** [comment]
I don't entirely agree that it would be that much code, but I like your solution. I will see what I can do with a release.sh script of my own. 😁

**Pull request closed without merging**